From: Jan Kara <jack@suse.cz>
Subject: Fix assertion failure in journal_write_metadata_buffer()
Patch-mainline: 2.6.18-rc1
References: 174813

Non-jbddirty buffers should be filed to BJ_Reserved and not BJ_Metadata list.
It can actually happen that we refile such buffers during the commit phase
when we reallocate in the running transaction blocks deleted in committing
transaction (and that can happen if the committing transaction already wrote
all the data and is just cleaning up BJ_Forget list).

Signed-off-by: Jan Kara <jack@suse.cz>

diff -rupX /home/jack/.kerndiffexclude linux-2.6.16-2-realloc_freed_fix/fs/jbd/transaction.c linux-2.6.16-3-refile_nodirty_fix/fs/jbd/transaction.c
--- linux-2.6.16-2-realloc_freed_fix/fs/jbd/transaction.c	2006-04-22 03:44:01.000000000 +0200
+++ linux-2.6.16-3-refile_nodirty_fix/fs/jbd/transaction.c	2006-05-25 01:58:14.000000000 +0200
@@ -2041,7 +2041,8 @@ void __journal_refile_buffer(struct jour
 	__journal_temp_unlink_buffer(jh);
 	jh->b_transaction = jh->b_next_transaction;
 	jh->b_next_transaction = NULL;
-	__journal_file_buffer(jh, jh->b_transaction, BJ_Metadata);
+	__journal_file_buffer(jh, jh->b_transaction,
+				was_dirty ? BJ_Metadata : BJ_Reserved);
 	J_ASSERT_JH(jh, jh->b_transaction->t_state == T_RUNNING);
 
 	if (was_dirty)
